sopel/tools/target.py

```diff
@@ -57,29 +67,35 @@ class Channel:
     """A representation of a channel Sopel is in.
 
     :param name: the channel name
-    :type name: :class:`~.tools.Identifier`
+    :type name: :class:`~sopel.tools.identifiers.Identifier`
+    :param identifier_factory: A factory to create
+                               :class:`~sopel.tools.identifiers.Identifier`
     """
-    def __init__(self, name):
-        assert isinstance(name, Identifier)
+    def __init__(
+        self,
+        name: identifiers.Identifier,
+        identifier_factory: IdentifierFactory = identifiers.Identifier,
+    ) -> None:
+        assert isinstance(name, identifiers.Identifier)
         self.name = name
         """The name of the channel."""
-        self.users = {}
+        self.users: Dict[identifiers.Identifier, User] = {}
         """The users in the channel.
 
-        This maps nickname :class:`~sopel.tools.Identifier`\\s to :class:`User`
-        objects.
+        This maps nickname :class:`~sopel.tools.identifiers.Identifier`\\s to
+        :class:`User` objects.
         """
-        self.privileges = {}
+        self.privileges: Dict[identifiers.Identifier, int] = {}
         """The permissions of the users in the channel.
 
-        This maps nickname :class:`~sopel.tools.Identifier`\\s to bitwise
-        integer values. This can be compared to appropriate constants from
-        :mod:`sopel.plugin`.
+        This maps nickname :class:`~sopel.tools.identifiers.Identifier`\\s to
+        bitwise integer values. This can be compared to appropriate constants
+        from :mod:`sopel.plugin`.
```

Comment:
As long as you're here?

```suggestion
        from :mod:`sopel.privileges`.
```

(Which now that I think about it, why isn't that `sopel.irc.privileges`? And why isn't the new location for `Identifier` in `sopel.irc.identifiers`? So many questions!)

Reply:
Long term planning: you can't move things around without causing cyclic import, thus breaking changes. I'm OK with that if you are, but I don't expect you to be (yet).